src/plugins/liveobjects/statemessage.ts

```diff
@@ -155,13 +155,13 @@ export class StateMessage {
 
     if (message.object?.map?.entries) {
       for (const entry of Object.values(message.object.map.entries)) {
-        decodeMapEntry(entry, inputContext, decodeDataFn);
+        await decodeMapEntry(entry, inputContext, decodeDataFn);
```

Comment:
is it possible to decode multiple map entries in parallel? if so then it will surely be way more efficient to build an array of promises and use `Promise.all`

Reply:
`decodeMapEntry` ends up calling the same [decodeData](https://github.com/ably/ably-js/blob/fbf9d422195aad85b291d1f46db010a9853742e9/src/common/lib/types/message.ts#L166) function for Message as we do when processing regular [channel messages](https://github.com/ably/ably-js/blob/fbf9d422195aad85b291d1f46db010a9853742e9/src/common/lib/client/realtimechannel.ts#L701) and [presence messages](https://github.com/ably/ably-js/blob/fbf9d422195aad85b291d1f46db010a9853742e9/src/common/lib/client/realtimechannel.ts#L605), and they do decoding sequentially.
there is only one possible async call in this case, and it's [cipher.decrypt](https://github.com/ably/ably-js/blob/fbf9d422195aad85b291d1f46db010a9853742e9/src/common/lib/types/message.ts#L218)

Reply:
it seems that we only support [AES in CBC mode](https://github.com/ably/ably-js/blob/fbf9d422195aad85b291d1f46db010a9853742e9/ably.d.ts#L764-L785), so I assume we can't decrypt concurrently?

Reply:
i'm definitely not an expert on cryptography, but afaict, while CBC mode doesn't allow parallel encryption, parallel decryption is perfectly fine

Reply:
i'd say we should change all 3 implementations to use `Promise.all`, worst that can happen is that the js engine processes them serially anyway.

either way, since it affects normal channel messages as well as liveobjects i'm happy for that to happen in another pr

Reply:
> parallel decryption is perfectly fine

yea, that's what google tells me too.
also, our `decrypt` [implementation ](https://github.com/ably/ably-js/blob/fbf9d422195aad85b291d1f46db010a9853742e9/src/platform/web/lib/util/crypto.ts#L223-L233) (included web only, but node.js is equivalent) doesn't seem to even be dependant on the previous ciphertext we received, as the iv is always included at the start of the ciphertext.

Reply:
at the same time, I don't really feel comfortable changing something like this as part of the LiveObjects work.
would you mine me creating a separate issue for this and asking people around if we can parallelize decryption in general? that would benefit regular and presence messages too.

Reply:
yep of course, sgtm 👍

Reply:
> either way, since it affects normal channel messages as well as liveobjects i'm happy for that to happen in another pr

yea, I would prefer that, will create an issue for that

Reply:
created https://github.com/ably/ably-js/issues/1906